### PR TITLE
[SPARK-32863][SS] Full outer stream-stream join

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

This PR is to add full outer stream-stream join, and the implementation of full outer join is:
* For left side input row, check if there's a match on right side state store.
  * if there's a match, output the joined row, o.w. output nothing. Put the row in left side state store.
* For right side input row, check if there's a match on left side state store.
  * if there's a match, output the joined row, o.w. output nothing. Put the row in right side state store.
* State store eviction: evict rows from left/right side state store below watermark, and output rows never matched before (a combination of left outer and right outer join).

### Why are the changes needed?

Enable more use cases for spark stream-stream join.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unit tests in `UnsupportedOperationChecker.scala` and `StreamingJoinSuite.scala`.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @viirya

**Comment by c21** [comment]
cc @HeartSaVioR and @xuanyuanking as well, thanks.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres @gaborgsomogyi as well

**Comment by HeartSaVioR** [comment]
Probably the general question would be this; "Is there a historical reason we didn't implement full outer join?"

The concept is straightforward so the code change should be also straightforward. Just wanted to know if there's specific reason not to implement stream-stream full outer join at the time when we implemented stream-stream left/right outer join, so that we could avoid missing any major considerations on this feature.

cc. to @tdas @jose-torres again here, as they've authored and reviewed the PR #19327

**Comment by c21** [comment]
Just to add my thought, and would like to get opinions from previous authors and reviewers in this area for sure.

* In terms of scalability, reliability, etc, full outer join stores same amount of data in state store, compared to left outer/right outer/inner join. The only difference is full outer join would output more rows (all rows from both stream sides). So I think from system perspective, we probably can support it, similar to other joins.

* The motivation from my side is that, I am investigating the performance difference between spark structured streaming (micro-batch) vs other internal streaming systems, and plan to add some more stuff on top of it to support internal use cases. So it would be good to have a basic full outer join implementation here (merge to upstream instead of forking for each spark version upgrade for me).

**Comment by c21** [comment]
Asked in dev mailing list as well in http://apache-spark-developers-list.1001551.n3.nabble.com/SS-full-outer-stream-stream-join-td30415.html to make sure we are not missing any voice.

**Comment by c21** [comment]
Shall we start review on this? It seems that no one raises concern here or mailing list. Would like to get this in before 3.1 deadline if this PR makes sense, thanks. cc @HeartSaVioR .

**Comment by xuanyuanking** [comment]
Spent some time on "Is there a historical reason we didn't implement full outer join?" but also don't know the answer.
Mind waiting for 1 more day since it's still the weekend for someone's side? +1 to get this in 3.1 if it makes sense.

**Comment by HeartSaVioR** [comment]
I'm sorry "blocking this" wasn't the main rationalization of requesting. My bad.

I'm OoO this week and don't have a time slot to review this change though - not sure I can handle new non-trivial reviews as I have another PRs to handle as well.

@viirya @xuanyuanking @gaborgsomogyi It'd be nice if you could review this. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 1]
```diff
@@ -438,4 +421,32 @@ object UnsupportedOperationChecker extends Logging {
     throw new AnalysisException(
       msg, operator.origin.line, operator.origin.startPosition, Some(operator))
   }
+
+  private def checkForStreamStreamJoinWatermark(join: Join): Unit = {
+    val watermarkInJoinKeys = StreamingJoinHelper.isWatermarkInJoinKeys(join)
+
+    val hasValidWatermarkRange = join.joinType match {
```
Comment (xuanyuanking) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala:428]: Let's also keep the comment https://github.com/apache/spark/pull/30395/files#diff-7c879c08d2f379c139d5229a88857229ae69bb48f0a138a3d64e1b2dde3502feL341 in this new method.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala [hunk 2]
```diff
@@ -382,6 +398,7 @@ case class StreamingSymmetricHashJoinExec(
             leftSideJoiner.removeOldState() ++ rightSideJoiner.removeOldState()
           case LeftOuter => rightSideJoiner.removeOldState()
           case RightOuter => leftSideJoiner.removeOldState()
+          case FullOuter => Iterator.empty
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:401]: Let's also add comments for FullOuter in: https://github.com/apache/spark/pull/30395/files#diff-6cd66da710d8d54025c1edf658bbec5230e8b4e748f9f2f884a60b1ba1efed42R395

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala [hunk 3]
```diff
@@ -333,13 +339,23 @@ case class StreamingSymmetricHashJoinExec(
...
         }.map(pair => joinedRow.withLeft(nullLeft).withRight(pair.value))
 
         hashJoinOutputIter ++ outerOutputIter
+      case FullOuter =>
+        lazy val isKeyToValuePairMatched = (kv: KeyToValuePair) =>
+          stateFormatVersion match {
+            case 2 => kv.matched
+            case _ => throwBadStateFormatVersionException()
+          }
+        val leftSideOutputIter = leftSideJoiner.removeOldState().filterNot(isKeyToValuePairMatched)
```
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:353]: Super nit, let's start the new line in the same position for `left/rightSideOutputIter`
```
val leftSideOutputIter = leftSideJoiner.removeOldState().filterNot(
  isKeyToValuePairMatched).map(pair => joinedRow.withLeft(pair.value).withRight(nullRight))
```

**Review by xuanyuanking — COMMENTED** [review]
Take a quick first-round review. I will run/review the tests tomorrow.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 4]
```diff
@@ -298,8 +298,14 @@ object UnsupportedOperationChecker extends Logging {
               // no further validations needed
 
             case FullOuter =>
-              if (left.isStreaming || right.isStreaming) {
-                throwError("Full outer joins with streaming DataFrames/Datasets are not supported")
+              if (left.isStreaming && !right.isStreaming) {
+                throwError(s"$FullOuter joins with streaming DataFrames/Datasets on the left " +
```
[the remark this answers is not in the document]
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala:302]: @xuanyuanking - both way works for me, changed to `FullOuter` here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 5]
```diff
@@ -315,40 +321,17 @@ object UnsupportedOperationChecker extends Logging {
...
-                }
+                checkForStreamStreamJoinWatermark(j)
               }
 
             // We support streaming right outer joins with static on the left always, and with
             // stream on both sides under the appropriate conditions.
             case RightOuter =>
               if (left.isStreaming && !right.isStreaming) {
-                throwError("Right outer join with a streaming DataFrame/Dataset on the left and " +
+                throwError(s"$RightOuter join with a streaming DataFrame/Dataset on the left and " +
```
[the remark this answers is not in the document]
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala:331]: @xuanyuanking - sure, updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala (hunk 1 above)
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala:428]: @xuanyuanking - added back. sorry I was missing that in the first place.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala (hunk 3 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:353]: @xuanyuanking - yeah I agree this is too nit, updated anyway.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:401]: @xuanyuanking - agree, updated.

**Comment by c21** [comment]
@xuanyuanking - addressed all existing comments and the PR is ready for review again, thanks.

**Review by HeartSaVioR — COMMENTED** [review]
Only comment regarding code comment. Otherwise looks OK to me.

**Comment by HeartSaVioR** [comment]
FYI, during reviewing I found confusing method names: `setupWindowedJoinWithRangeCondition` / `setupWindowedSelfJoin`. We should remove `Windowed` there, but not from this PR as these methods exist. 

@c21 Would you mind submitting a new MINOR PR for this? Otherwise I'll do it instead.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 6]
```diff
@@ -411,11 +411,12 @@ class UnsupportedOperationsSuite extends SparkFunSuite with SQLHelper {
 
   // Full outer joins: only batch-batch is allowed
```
[the remark this answers is not in the document]
Comment (c21) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:412]: I agree this is a bit confusing, and I was confused when I looked at the unit test earlier. The test (`testBinaryOperationInStreamingPlan `) is to check whether `_.join(_, joinType = _)` supports stream-stream or not. For left outer/right outer/left semi/full outer, stream-stream join is not supported in this case as the watermark condition is not defined. See original unit tests for left outer/right outer/left semi [here](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala#L426-L435).

**Comment by c21** [comment]
> FYI, during reviewing I found confusing method names: setupWindowedJoinWithRangeCondition / setupWindowedSelfJoin. We should remove Windowed there, but not from this PR as these methods exist.

@HeartSaVioR - sounds good, I will submit a separate PR later tomorrow. thanks.

**Comment by c21** [comment]
@xuanyuanking , @HeartSaVioR - wondering do you have any more comment? Thanks.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:412]: The comment was explaining for "all cases" of full outer joins. I know the test is covered but the valid comment becomes no longer valid and we still leave the comment here.

I meant like below:

```
// Full outer joins: stream-batch/batch-stream join are not allowed
// stream-stream full outer join is allowed 'conditionally' - see below check
```

**Review by HeartSaVioR — APPROVED** [review]
+1

I'll leave this in a couple of days to see further comments and merge if there's no new input. Thanks.

**Review by xuanyuanking — APPROVED** [review]
LGTM, thanks for the work!

**Comment by HeartSaVioR** [comment]
All current reviewers are OK with this, so I'm kicking the Github Action build again. Once the test passes I'll merge this.

**Review by viirya — COMMENTED** [review]
Not looking closely at the test part, the hash join code looks okay.

**Comment by c21** [comment]
I am seeing the unit test failure `org.apache.spark.storage.BlockManagerDecommissionIntegrationSuite` from https://github.com/apache/spark/pull/30395/checks?check_run_id=1478597516, I don't think the failed one is relevant. @HeartSaVioR - could you help re-trigger the tests? Thanks.

**Comment by HeartSaVioR** [comment]
@viirya Do you have a plan to review this? Then I'll open this for a day more (as actually planned).

**Comment by HeartSaVioR** [comment]
@c21 Jenkins is down for a couple of days, so we have to rely on Github Action. Closing and reopening PR would work for you.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 7]
```diff
@@ -409,13 +409,15 @@ class UnsupportedOperationsSuite extends SparkFunSuite with SQLHelper {
...
     expectedMsg = "is not supported in Update output mode")
 
-  // Full outer joins: only batch-batch is allowed
+  // Full outer joins: stream-batch/batch-stream join are not allowed,
+  // and stream-stream join is allowed 'conditionally' - see below check
   testBinaryOperationInStreamingPlan(
-    "full outer join",
+    "FullOuter join",
     _.join(_, joinType = FullOuter),
     streamStreamSupported = false,
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:417]: streamStreamSupported should be true now?

Comment (c21) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:417]: Please see above discussion - https://github.com/apache/spark/pull/30395#discussion_r532009689 . thanks.

**Review by viirya — APPROVED** [review]
One remaining comment. Otherwise looks okay.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala (hunk 7 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala:417]: Ok.

**Comment by c21** [comment]
@viirya - wondering is my reply above addressing your comment? Thanks.

**Comment by viirya** [comment]
> @viirya - wondering is my reply above addressing your comment? Thanks.

Yeah, I didn't see your reply before I submitted the review comment.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Jenkins doesn't seem to work, or struggling to do with queued builds. I'll trigger Github Action instead, and merge once it passes.

**Comment by HeartSaVioR** [comment]
OK test passed. Merging to master.

**Comment by c21** [comment]
Thanks @HeartSaVioR , @xuanyuanking and @viirya for review, and making it into spark 3.1 !

**Pull request closed without merging**